grpc-sys/build.rs

```diff
@@ -39,133 +38,155 @@ mod imp {
-            .build();
 
-        println!("cargo:rustc-link-search=native={}/build",
-                 zlib_dst.display());
+    pub fn build_or_link_grpc(cc: &mut ::gcc::Config) {
+        let out_dir = env::var("OUT_DIR").expect("Can't access OUT_DIR");
+        DirBuilder::new()
+            .recursive(true)
+            .create(format!("{}/{}", out_dir, "grpc"))
+            .unwrap();
+
+        FakeGitSubmodule::new_with_semver("grpc", "grpc", GRPC_VERSION, "grpc")
+            .init()
+            .unwrap();
+
+        FakeGitSubmodule::new_with_semver("madler", "zlib", ZLIB_VERSION, "grpc/third_party/zlib")
+            .init()
+            .unwrap();
+
+        FakeGitSubmodule::new_with_hash("google",
+                                        "boringssl",
+                                        BORINGSSL_COMMIT_HASH,
+                                        "grpc/third_party/boringssl")
+                .init()
+                .unwrap();
 
-        let cares_dst = ::cmake::Config::new(format!("grpc-{}/src/c-ares", GRPC_VERSION))
-            .out_dir(format!("{}/{}", out_dir, "cares"))
-            .build_target("cares")
-            .build();
+        FakeGitSubmodule::new_with_tag("c-ares",
+                                       "c-ares",
+                                       CARES_TAG,
+                                       "grpc/third_party/cares/cares")
+                .init()
+                .unwrap();
 
-        println!("cargo:rustc-link-search=native={}/build",
-                 cares_dst.display());
+        // fix multiple _main symbols
+        let _ = Command::new("rm")
+            .current_dir(&format!("{}/grpc/third_party/cares/cares", out_dir))
+            .args(&["acountry.c", "adig.c", "ahost.c"])
+            .status();
 
-        let dst = ::cmake::Config::new(format!("grpc-{}", GRPC_VERSION))
-            .define("ZLIB_ROOT_DIR",
-                    format!("{}/zlib-{}", crate_dir, ZLIB_VERSION))
-            .define("BORINGSSL_ROOT_DIR",
-                    format!("{}/boringssl-{}", crate_dir, BORINGSSL_GIT_HASH))
-            .out_dir(format!("{}/{}", out_dir, "grpc"))
-            .build_target("grpc_unsecure")
+        let dst = ::cmake::Config::new(format!("{}/grpc", out_dir))
+            .build_target("grpc")
             .build();
 
         println!("cargo:rustc-link-search=native={}/build", dst.display());
+        println!("cargo:rustc-link-search=native={}/build/third_party/cares",
+                 dst.display());
+        println!("cargo:rustc-link-search=native={}/build/third_party/zlib",
+                 dst.display());
+        println!("cargo:rustc-link-search=native={}/build", dst.display());
 
         println!("cargo:rustc-link-lib=static=z");
         println!("cargo:rustc-link-lib=static=cares");
         println!("cargo:rustc-link-lib=static=gpr");
-        println!("cargo:rustc-link-lib=static=grpc_unsecure");
+        println!("cargo:rustc-link-lib=static=grpc");
```

Comment:
grpc uses its own boring ssl as ssl implementation.
haven't link against ssl/crypto. ssl may not works.
here `grpc` for simplifying compile instructions. (or else one must compile cares, z, etc one by one)